pkg/identity/cache/local.go

```diff
@@ -57,18 +68,32 @@ func (l *localIdentityCache) getNextFreeNumericIdentity(idCandidate identity.Num
 			// let nextNumericIdentity be, allocated identities will be skipped anyway
 			log.Debugf("Reallocated restored local identity: %d", idCandidate)
 			return idCandidate, nil
+		} else {
+			log.WithField(logfields.Identity, idCandidate).Debug("Requested local identity not available to allocate")
 		}
 	}
 	firstID := l.nextNumericIdentity
 	for {
 		idCandidate = l.nextNumericIdentity | l.scope
-		if _, taken := l.identitiesByID[idCandidate]; !taken {
+		_, taken := l.identitiesByID[idCandidate]
+		_, reserved := l.reservedIdentities[idCandidate]
+		if !taken && !reserved {
 			l.bumpNextNumericIdentity()
 			return idCandidate, nil
 		}
 
 		l.bumpNextNumericIdentity()
 		if l.nextNumericIdentity == firstID {
+			// Desperation: no local identities left (unlikely). If there are reserved
+			// but not-taken identities, claim one of them.
+			for reservedID := range l.reservedIdentities {
+				if _, taken := l.identitiesByID[reservedID]; !taken {
+					delete(l.reservedIdentities, reservedID)
+					log.WithField(logfields.Identity, reservedID).Warn("Local identity allocator full; claiming first reserved identity. This may cause momentary policy drops")
+					return reservedID, nil
+				}
+			}
+
```

Comment:
Do we need this case?

I'm a little concerned that in this case we're in a scenario of "no resources left" and the tradeoff may end up be between "impact existing traffic" or "impact new traffic", and this extra logic might end up tipping us over into "impact existing traffic" territory. The warning is useful to understand we're in this situation, but I struggle a little bit to decide whether it's a clear win to have this particular behaviour in that scenario.

I'd be interested in any insights you might have around the background thinking to this  logic.

Reply:
Maybe to expand some perspectives for discussions' sake:

- I can understand that if for some reason users of this code hold their reservations infinitely, this provides an escape hatch. However, that is a bug in the caller APIs, so we should be testing & validating usage of the API rather than solving it here (?)
- Consider a case where the set of identities to reserve is exceedingly high. During startup it would first reserve many identities. Over time k8s resources will get synced and may also attempt to allocate identities. In an ideal situation I think that the restore logic should restore the identities, trigger the usage of them, then remove all reservations, and ideally afterwards all other k8s resource sync logic would then start to allocate identities from the remaining space. This would have the smallest impact on the changes to identities during restore, and hence the lowest likelihood of connection impact. On the other hand it also assumes that whatever other allocation logic that exists would be able to retry in order to eventually allocate an identity from the actual available remaining allocation space. Let's say that this code continues to exist, it could mean that some of the restored identities will transition their identities unnecessarily, and in the worst case in ways that may not be able to be recovered from.

In some sense when you're at/near the limit maybe things are borked anyway, but it's one thing to say that "there are no more identities"; it's another thing to say "there are no more identities and by the way Cilium decided to arbitrarily rearrange which peers have identities now".

Reply:
I thought about this for a while, and decided that allowing allocations to make progress was the better choice of the two. The reason is, basically:
- If the ipcache can't make progress, then we have less of a chance of releasing identities to free up room
- If local allocation is failing, then we're potentially causing policy drops in the case of FQDN policies
- Identity reservation is best-effort anyways :-)
- The most likely scenario where we hit this would be a bug with identity restoration, in which case a policy flap is better than a livelock

That is an interesting point about shortening the grace period -- we probably can just wait for a full kvstore / k8s sync and be done after that. That lessens the time where identity pressure would be exacerbated by reservations.

FWIW, once #28652 or something similar is merged, there should be no policy drops if identities do flap on restart.

Reply:
Hmm. So we're debating the case of behaviour when allocating (16M - N reserved) local identities right? When I phrase it that way it seems like such a corner case that we're unlikely to hit this anyway, so it's barely worth debating. Probably my key principle for such a rare case would be just to make it fail as simply as possible, no smarts, just make the behaviour obvious. To be fair, the log statement is good enough to check that box.

With high N I agree that it's useful to minimize the risk of some livelock type scenario where the total identity space is big enough but somehow we can't allocate because N gets out of control. I suppose my argument was thinking more like "I wouldn't want to debug a system that ignores the reservations and hence rearranges the prefix <-> identity mappings upon restart." That said, I'm happy to defer to you on this one. Between the rarity, the warning log and the test cases, I think it's good enough.